font/src/ft/mod.rs

```diff
@@ -529,6 +589,86 @@ impl FreeTypeRasterizer {
     }
 }
 
+fn downsample_bitmap(mut bitmap_glyph: RasterizedGlyph, fixup_factor: f64) -> RasterizedGlyph {
+    // Don't try to upscale
+    if fixup_factor > 1.0 {
+        return bitmap_glyph;
+    }
+
+    let bitmap_width = bitmap_glyph.width as f64;
+    let bitmap_height = bitmap_glyph.height as f64;
+
+    let width = (bitmap_width * fixup_factor) as usize;
+    let height = (bitmap_height * fixup_factor) as usize;
+
+    let bitmap_width = bitmap_width as usize;
+    let bitmap_height = bitmap_height as usize;
+
+    let b_buf = &bitmap_glyph.buf;
+    let scaling_factor =
+        (bitmap_width as f32 / width as f32).max(bitmap_height as f32 / height as f32);
+    let advance_step = scaling_factor.ceil() as usize;
+    let mut scaled_buffer = Vec::with_capacity(width * height * 3);
+
+    let mut new_line_index = 0;
+    let mut source_line_index = 0;
+
+    while new_line_index < height {
+        let mut new_column_index = 0;
+        let mut source_column_index = 0;
+
+        while new_column_index < width {
+            let mut r: u32 = 0;
+            let mut g: u32 = 0;
+            let mut b: u32 = 0;
+            let mut pixels_picked: u32 = 0;
+
+            let source_end_line = std::cmp::min(source_line_index + advance_step, bitmap_height);
+            let source_end_column = std::cmp::min(source_column_index + advance_step, bitmap_width);
+
+            let mut source_line_index = source_line_index;
+            while source_line_index < source_end_line {
+                let cur_pixel_index = source_line_index * bitmap_width * 3;
+
+                let mut source_column_index = source_column_index;
+                while source_column_index < source_end_column {
```

Comment:
```suggestion
                for i in source_column_index..source_end_column {
```

We definitely do not want to shadow `source_column_index` and `source_line_index`. Especially in the way it is right now, the `source_column_index += 1` could easily explode right in our face.

Reply:
How it could explode?

Reply:
Shadowing will cause problems when you refactor code, say you want to change to my `for` loop, so you delete the original `let mut variable` statement. The compiler will not warn  you about removing the `variable += 1`, since the previous variable was shadowed, thus causing a bug that could have easily been prevented.

So in this case variable shadowing is a really bad idea, because by moving around or deleting lines the code easily breaks in hard to understand ways.

If you do things like shadow variables of a different type like `let path = path.to_string()`, then you don't run into these problems, due to help from the typesystem.